docs/contributing/code_guide.rst

```diff
@@ -279,39 +279,138 @@ to edit the pull request by changing the corresponding branch on your
+
+  .. code-block:: python
+
+     collision_frequency = 1e6 * u.s ** -1
+
+* Use comments to communicate information that you wish you knew before
+  starting to work on a particular section of code, including
+  information that took some time to learn.
+
+* Use comments to communicate information that the code cannot,
+  such as why an alternative approach was *not* taken.
+
+* When possible, include enough contextual information that the comment
+  will make sense even if it is displaced from the code it was
+  originally describing.
+
+* Use comments to include references to books or articles that describe
+  the equation, algorithm, or software design pattern that is being
+  implemented. Even better, include these references in docstrings.
+
+* Provide enough contextual information in the comment for a new user
+  to be able to understand it.
+
+* Remove commented out code before merging a pull request.
+
+* When updating code, be sure to update the comments too!
+
+* When a comment is used as the header for a section of code, consider
+  extracting that section of code into its own function. For example, we
+  might start out with a function that includes multiple lines of code
+  for each step.
+
+  .. code-block:: python
+
+     def analyze_experiment(data):
+         # Step 1: calibrate the data
+         ...
+         # Step 2: normalize the data
+         ...
+
+  We can apply the `extract function refactoring pattern`_ by creating a
+  separate function for each of these steps. The name of each function
+  can often be extracted directly from the comment.
+
+  .. code-block:: python
+
+     def calibrate_data(data):
+         ...
+         return calibrated_data
+
+     def normalize_data(data):
+         ...
+         return normalized_data
+
+     def analyze_experiment(data):
+         calibrated_data = calibrate_data(data)
+         normalized_data = normalize_data(calibrated_data)
+
+  This refactoring strategy is appropriate for long functions where the
+  different steps can be cleanly separated from each other. This pattern
+  leads to functions that are shorter, more focused, more reusable, and
+  easier to test. The original function no longer includes
+  implementation details, and thus gives a high level view of what the
+  function is doing.
+
+  .. hint::
+
+     The `extract function refactoring pattern`_ might not be
+     appropriate if the different sections of code are intertwined with
+     each other (e.g., if both sections require the same intermediate
+     variables). An alternative in such cases would be to create a class
+     instead.
 
-* Separate subject from body with a blank line
+Error messages
+==============
 
-* Limit the subject line to 50 characters
+Error messages are a vital but underappreciated form of documentation.
+A good error message can help someone pinpoint the source of a problem
+in seconds, while a cryptic or missing error message can lead to hours
+of frustration.
```

Comment:
I had a `sphinx` error like this yesterday.  It didn't even tell me what file the error was in. Ahhhh... 😤

Reply:
YEP.